infrastructure/src/main/java/org/corfudb/infrastructure/RequestHandlerMethods.java

```diff
@@ -0,0 +1,154 @@
+
+            HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getUnknownErrorMsg(e)), ctx);
+        }
+    }
+
+    /** Generate handlers for a particular server.
+     *
+     * @param caller    The context that is being used. Call MethodHandles.lookup() to obtain.
+     * @param server    The object that implements the server.
+     * @return          New request handlers for caller class.
+     */
+    public static RequestHandlerMethods generateHandler(@Nonnull final MethodHandles.Lookup caller,
+                                                        @NonNull final AbstractServer server) {
+        RequestHandlerMethods handler = new RequestHandlerMethods();
+        Arrays.stream(server.getClass().getDeclaredMethods())
+                .filter(method -> method.isAnnotationPresent(RequestHandler.class))
+                .forEach(method -> handler.registerMethod(caller, server, method));
+        return handler;
+    }
+
+    private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
+                                @Nonnull final AbstractServer server,
+                                @Nonnull final Method method) {
+        final RequestHandler annotation = method.getAnnotation(RequestHandler.class);
+
+        if(handlerMap.containsKey(annotation.type())) {
+            throw new UnrecoverableCorfuError("HandlerMethod for " + annotation.type() + " already registered!");
+        }
+
+        try {
+            HandlerMethod h;
+            if (Modifier.isStatic(method.getModifiers())) {
+                MethodHandle mh = caller.unreflect(method);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller,
+                        "handle", MethodType.methodType(HandlerMethod.class),
+                        mh.type(), mh, mh.type()).getTarget().invokeExact();
+            } else {
+                // Instance method, so we need to capture the type.
+                MethodType mt = MethodType.methodType(method.getReturnType(), method.getParameterTypes());
+                MethodHandle mh = caller.findVirtual(server.getClass(), method.getName(), mt);
+                MethodType mtt = mh.type().dropParameterTypes(0, 1);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller, "handle",
+                        MethodType.methodType(HandlerMethod.class, server.getClass()),
+                        mtt, mh, mtt).getTarget().bindTo(server).invoke();
+            }
+
+            // Install pre-conditions on handler and place the handler in the map
+            final HandlerMethod handler = generateConditionalHandler(annotation.type(), h);
+            handlerMap.put(annotation.type(), handler);
+        } catch(Throwable e) {
+            log.error("registerMethod: Exception during request handler registration", e);
+            throw new UnrecoverableCorfuError(e);
```

Comment:
Wonder if we can embed the error message into the error itself?
It's generally preferable to not log and then throw an exception.

Reply:
Done.